[docs] Add a Linking section (#30958)
expo/expo: docs/pages/linking/overview.mdx

---
title: Overview of Linking, Deep Links, Android App Links, and iOS Universal Links
sidebar_title: Overview
description: An overview of available resources to implement Linking and Deep Links in your Expo apps.
---

import { AndroidIcon } from '@expo/styleguide-icons/custom/AndroidIcon';
import { AppleIcon } from '@expo/styleguide-icons/custom/AppleIcon';
import { BookOpen02Icon } from '@expo/styleguide-icons/outline/BookOpen02Icon';

import { BoxLink } from '~/ui/components/BoxLink';
import { CODE } from '~/ui/components/Text';

## Linking

Linking allows your app to interact with incoming and outgoing URLs. In this process, the user not only gets directed to open your app, but they are taken to a specific screen (route) within the app.

For example, by clicking an advertisement of your product, your app will open on the user's device and they can view that product's details. This product's link that the user clicked may look like:

> ```html

Reviewer: See comment below; this is actually a deep link, I believe.

Author: I've moved this example to Deep links section

myapp://web-app.com/product
```

This link is constructed by three parts:

- **Scheme**: The URL scheme that identifies the app that should open the URL (example: `myapp://`). It can also be `https` or `http` for non-standard deep links.
- **Host**: The domain name of the app that should open the URL (example: `web-app.com`).
- **Path**: The path to the screen that should be opened (example: `/product`). If the path isn't specified, the user is taken to the home screen of the app.

## Linking strategies

There are different linking strategies you handle in your Expo app:

> - Linking to other apps from your app (outgoing links)

Reviewer: I might organize these differently to reinforce the differences between linking, deep linking, and the similarities between app links and universal links:
```
- Linking to other apps from your app (outgoing links)
- Linking to your app from other apps or a web browser (incoming links)
- Linking to specific content within your app (deep links)
- Redirecting links to your web domain to your app if it is installed (Android App Links and iOS Universal Links)
```

- Linking to your app from other apps or a web browser (incoming links)
- Android App Links
- iOS Universal Links

> **info** **Tip:** We recommend using [Development builds](/develop/development-builds/introduction/) to test your app's linking strategies instead of Expo Go.

### Into other apps

Linking to other apps from your app is achieved using a URL based on the target app's URL scheme. This **URL scheme** allows you to reference resources within that native app.

Your app can use a [common URL scheme](/linking/into-other-apps/#common-url-schemes) for default apps or implement any URL scheme, including `https` and `http` (commonly used by web browsers like Chrome, Safari, and so on), and use JavaScript to handle the URL that launches the corresponding native app.

<BoxLink
  title="Linking into other apps"
  description="Learn how to handle common and custom URL schemes to link other apps from your app."
  href="/linking/into-other-apps/"
  Icon={BookOpen02Icon}
/>

### Into your app

> Linking to your app from other apps or a web browser is achieved using a URL based on the scheme for your app. These URLs are called **deep links**.

Reviewer: The document introduces "links" and "deep links" in the title. This is the first time afterwards where it references "deep links". If we're presenting these as two concepts (which I think is useful on an overview page like this), let's differentiate them more strongly. I think this definition is a good one to anchor this to: https://en.wikipedia.org/wiki/Deep_linking. Deep links aren't specific to apps; they're just links to specific url-based content inside of a website or app. 

Thinking back to apps, any app with a registered scheme can be linked to (e.g., `myapp://whatever`). Only apps that listen for deep links and perform routing based on them really implement "deep linking", and it just happens that Expo Router makes this happen automatically. (seems like long ago, but a lot of web apps used to not support deep linking... actually, many still don't).

<BoxLink
  title="Linking to your app"
  description="Learn how to configure custom URL schemes to create a deep link of your app."
  href="/linking/into-your-app/"
  Icon={BookOpen02Icon}
/>

> ### Android App Links

Reviewer: Can we group app links and universal links under a single heading, and then break out in much more brief into each platform? The story for each is basically the same: "This is how you route an http(s) link to a particular domain to your app if it is installed", just they're named differently per platform. I liken these to "App Shortcuts / Quick Actions": same concept that we'd probably call by the same name if Google and Apple could settle on one, and even then, we probably just pick the name from one platform and use it colloquially to refer to both.

Author: We used to use Universal links before this PR made a division. Do you want me to still use "Universal links"?

Author: Here's the update based on your initial feedback:

![CleanShot 2024-09-23 at 19 27 20](https://github.com/user-attachments/assets/14b9056f-b301-4f32-a891-91ec043221ab)

<BoxLink
  title="Configure Android App Links"
  description={
    <>
      Learn how to configure <CODE>intentFilters</CODE> and set up two-way association from a
      standard web URL.
    </>
  }
  href="/linking/android-app-links/"
  Icon={AndroidIcon}
/>

### iOS Universal Links

iOS Universal Links are different from [standard deep links](#into-your-app) as they use regular HTTP and HTTPS schemes and are exclusive to iOS devices.

This link type allows your app to always open when a user clicks the link instead of choosing between the browser or another handler from a dialog displayed on the device. If the user doesn't have your app installed, the link takes them to your app's associated website. You can further configure the website by displaying a banner to the user to open your app using [Apple Smart Banner](/linking/ios-universal-links/#apple-smart-banner).

<BoxLink
  title="Configure iOS Universal Links"
  description={
    <>
      Learn how to configure <CODE>associatedDomains</CODE> and set up two-way association.
    </>
  }
  href="/linking/ios-universal-links/"
  Icon={AppleIcon}
/>

## Use Expo Router to handle deep linking

> To implement any of the above Linking strategy, **we recommend using** [Expo Router](/router/introduction/) since deep linking is automatically enabled for all of your app's screens.

Reviewer: ```suggestion
To implement any of the above Linking strategies, **we recommend using** [Expo Router](/router/introduction/) since deep linking is automatically enabled for all of your app's screens.
```

**Benefits:**

> - When implementing deep links, you can skip [adding `scheme` to your app's config](/linking/into-your-app/#add-a-custom-scheme-in-app-config).

Reviewer: I'm fuzzy about how this is a benefit of Expo Router. All Expo apps register the bundle ID as a default scheme, I believe. Router helps define the url after the scheme, but doesn't affect the scheme itself, AFAIK.

Author: I'll remove it from here but on testing this, I found if you are using Expo Router, there's no requirement to use `scheme` and instead, an app can use ios bundler identifier or android package name to create deep links.

- `Link` component from Expo Router can be used to [handle URL schemes to other apps](/linking/into-other-apps/#expo-router)

> - Android App Links and iOS Universal Links require configuring two-way association and runtime routing for the link in your app. This configuration is done in JavaScript and must be applied to every route to ensure coordination between web and native platforms. Since Expo Router enables deep links automatically, you don't have to configure runtime routing separately.

Reviewer: Also fuzzy on this being a router-specific benefit ... Router requires the same universal link configuration in order for universal links to work. I guess one indirect benefit is that if your bespoke non-router link handling matched on the scheme as well as the deep link, that would break with a universal link (or any other scheme, for that matter).

Author: This piece of content was already there in both sections (App and Universal links) as a note and was presented as a benefit of Expo Router. Instead of repeating ourselves in both documents, I decided to add this here as one place to gather all the info.

Based on your feedback, should I re-word this to say "hey, this is an indirect benefit of using Expo Route". On my research, I was also comparing this React navigation where you have to define a prefix: https://reactnavigation.org/docs/deep-linking/. iiuc, this is handled by Expo Router automatically.

  href="/linking/into-other-apps/"
  Icon={BookOpen02Icon}
/>

## Linking to your app from other apps or a web browser

[Deep Links](https://en.wikipedia.org/wiki/Deep_linking) are links to a specific URL based content inside an app or a website.

For example, by clicking an advertisement of your product, your app will open on the user's device and they can view that product's details. This product's link that the user clicked may look like:

```html
myapp://web-app.com/product
```

This link is constructed by three parts:

- **Scheme**: The URL scheme that identifies the app that should open the URL (example: `myapp://`). It can also be `https` or `http` for non-standard deep links.
- **Host**: The domain name of the app that should open the URL (example: `web-app.com`).
- **Path**: The path to the screen that should be opened (example: `/product`). If the path isn't specified, the user is taken to the home screen of the app.

<BoxLink
  title="Linking to your app"
  description="Learn how to configure custom URL schemes to create a deep link of your app."
  href="/linking/into-your-app/"
  Icon={BookOpen02Icon}
/>

> ## App shortcuts

Reviewer: ```suggestion
## Universal linking

Both Android and iOS implement their own systems for routing web URL's to an app if the app is installed. On Android, this system is called App Links, and on iOS it is called Universal Links.
```
Maybe something like this? I do tend to see this referred to as "universal links" or "universal linking" colloquially.

Author: Yeah, we used to call the group "universal links" before. Sounds good to me!

import { BookOpen02Icon } from '@expo/styleguide-icons/outline/BookOpen02Icon';

import { BoxLink } from '~/ui/components/BoxLink';
import { CODE } from '~/ui/components/Text';

## Linking

Linking allows your app to interact with incoming and outgoing URLs. In this process, the user not only gets directed to open your app, but they are taken to a specific screen (route) within the app.

### Linking strategies

There are different linking strategies you handle in your Expo app:

- Linking to other apps from your app (outgoing links)
- Linking to your app from other apps or a web browser (incoming links)
- Linking to specific content within your app (deep links)
- Redirecting links to your web domain to your app if it is installed (Android App Links and iOS Universal Links); also known as App shortcuts

> **info** **Tip:** We recommend using [Development builds](/develop/development-builds/introduction/) to test your app's linking strategies instead of Expo Go.

## Linking to other apps from your app

Linking to other apps from your app is achieved using a URL based on the target app's URL scheme. This **URL scheme** allows you to reference resources within that native app.

> Your app can use a [common URL scheme](/linking/into-other-apps/#common-url-schemes) for default apps or implement any URL scheme, including `https` and `http` (commonly used by web browsers like Chrome, Safari, and so on), and use JavaScript to handle the URL that launches the corresponding native app.

Author: You are right, yes invoking the URL is the action that the user can take here by following the link below.

  href="/linking/into-your-app/"
  Icon={BookOpen02Icon}
/>

## Universal linking

Both Android and iOS implement their own systems for routing web URL's to an app if the app is installed. On Android, this system is called App Links, and on iOS it is called Universal Links.

### Android App Links

Android App Links are different from [standard deep links](#into-your-app) as they use regular HTTP and HTTPS schemes and are exclusive to Android devices.

This link type allows your app to always open when a user clicks the link instead of choosing between the browser or another handler from a dialog displayed on the device. If the user doesn't have your app installed, the link takes them to your app's associated website.

<BoxLink
  title="Configure Android App Links"
  description={
    <>
      Learn how to configure <CODE>intentFilters</CODE> and set up two-way association from a
      standard web URL.
    </>
  }
  href="/linking/android-app-links/"
  Icon={AndroidIcon}
/>

### iOS Universal Links

iOS Universal Links are different from [standard deep links](#into-your-app) as they use regular HTTP and HTTPS schemes and are exclusive to iOS devices.

> This link type allows your app to always open when a user clicks the link instead of choosing between the browser or another handler from a dialog displayed on the device. If the user doesn't have your app installed, the link takes them to your app's associated website. You can further configure the website by displaying a banner to the user to open your app using [Apple Smart Banner](/linking/ios-universal-links/#apple-smart-banner).

Reviewer: ```suggestion
This link type allows your app to open when a user clicks an HTTP(S) link pointing to your web domain. If the user doesn't have your app installed, the link takes them to your app's associated website. You can further configure the website by displaying a banner for the user to open your app using [Apple Smart Banner](/linking/ios-universal-links/#apple-smart-banner).
```
I don't use iOS as much, but I don't recall iOS prompting for how to handle a link. Usually one app clobbers another if they're sitting on the same scheme, and I thought that an https link would just go to the browser if there's no universal linking app registered

Author: Yes you are right. On IOS, they will be directed to the website. I think I had the wrong context here.
